### PR TITLE
Enable tenant ID selection for multi-tenant queries.

### LINKED ISSUE #5820: Support selections on `__tenant_id__` in multi-tenant queries.
**Is your feature request related to a problem? Please describe.**
So far multi-tenant queries results include the `__tenant_id__` label. However, it is not possible to filter on them with e.g. `{__tenant_id__="1"}`.

**Describe the solution you'd like**
The filters should be supported.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:
This pull request will add tenant ID filtering such as `{__tenant_id__="1"}` to multi-tenant queries. It does not add filter such as `{app="loki"} | __tenant_id__="1"`.

We are parsing the select logs or select samples queries and find the matcher. It is then removed and the non-matching tenants are ignored.

**Which issue(s) this PR fixes**:
Fixes #5820

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [x] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/querier/multi_tenant_querier.go [hunk 1]
```diff
@@ -65,15 +74,24 @@ func (q *MultiTenantQuerier) SelectSamples(ctx context.Context, params logql.Sel
 		return q.Querier.SelectSamples(ctx, params)
 	}
 
+	selector, err := params.LogSelector()
+	if err != nil {
+		return nil, err
+	}
+	matchedTenants, filteredMatchers := filterValuesByMatchers(defaultTenantLabel, tenantIDs, selector.Matchers()...)
+	params.Selector = selector.WithMatchers(filteredMatchers).String() // TODO: this is wrong.
```
Comment (jeschkies) [inline pkg/querier/multi_tenant_querier.go:82]: I need to modify the complete expression not just the selector.

File: pkg/querier/multi_tenant_querier.go [hunk 2]
```diff
@@ -136,6 +154,57 @@ func (q *MultiTenantQuerier) Series(ctx context.Context, req *logproto.SeriesReq
 	return logproto.MergeSeriesResponses(responses)
 }
 
+// See https://github.com/grafana/mimir/blob/114ab88b50638a2047e2ca2a60640f6ca6fe8c17/pkg/querier/tenantfederation/tenant_federation.go#L29-L69
```
Comment (jeschkies) [inline pkg/querier/multi_tenant_querier.go:157]: Should we share this code or is copying fine?

File: pkg/logql/syntax/ast.go [hunk 3]
```diff
@@ -37,6 +37,7 @@ func (implicit) logQLExpr() {}
 // LogSelectorExpr is a LogQL expression filtering and returning logs.
 type LogSelectorExpr interface {
 	Matchers() []*labels.Matcher
+	WithMatchers([]*labels.Matcher) LogSelectorExpr
```
Comment (jeschkies) [inline pkg/logql/syntax/ast.go:40]: I wonder if this is some scope creep. The sharding mappers use `Walk` instead. Should we open the door for AST manipulation here?

File: pkg/querier/multi_tenant_querier.go [hunk 4]
```diff
@@ -136,6 +153,94 @@ func (q *MultiTenantQuerier) Series(ctx context.Context, req *logproto.SeriesReq
...
+	if err != nil {
+		return nil, nil, err
+	}
+	matchedTenants, filteredMatchers := filterValuesByMatchers(defaultTenantLabel, tenantIDs, expr.Selector().Matchers()...)
+	expr = replaceMatchers(expr, filteredMatchers)
+	return matchedTenants, expr, nil
+}
+
+// replaceMatchers traverses the passed epxression and replaces all matchers.
+func replaceMatchers(expr syntax.SampleExpr, matchers []*labels.Matcher) syntax.SampleExpr {
```
Comment (jeschkies) [inline pkg/querier/multi_tenant_querier.go:168]: @ssncferreira and @chaudum, I've added you because I saw a mapper in your recent changes. I'm wondering if we should add a visitor pattern to the AST. What do you think?

**Comment by jeschkies** [comment]
It works :tada: 
image

**Review by cyriltovena — APPROVED** [review]
LGTM

just improve the LogQL parsing.

**Review by JordanRushing — APPROVED** [review]
Looks great! I've drafted a quick test for the `sliceToSet()` util function: https://github.com/JordanRushing/loki/commit/8922c29112322f596d6711456e85fe73895c361a -- feel free to bring it in to this PR but not blocking.
